README.md
<h2 align="center">PMAW: Pushshift Multithread API Wrapper</h2>

[![CircleCI](https://circleci.com/gh/mattpodolak/pmaw.svg?style=shield)](https://circleci.com/gh/mattpodolak/pmaw)
[![codecov.io](https://codecov.io/github/mattpodolak/pmaw/coverage.svg?branch=master)](https://codecov.io/github/mattpodolak/pmaw)
[![PyPI Version](https://img.shields.io/pypi/v/pmaw?color=blue)](https://pypi.org/project/pmaw/)
[![Python Version](https://img.shields.io/pypi/pyversions/pmaw?color=blue)](https://www.python.org/)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

# Contents

- [Description](#description)
- [Getting Started](#getting-started)
- [Features](#features)
  - [Multithreading](#multithreading)
  - [Rate Limiting](#rate-limiting)
  - [Caching](#caching)
  - [PRAW Enrichment](#praw-enrichment)
  - [Custom Filtering](#custom-filtering)
  - [Unsupported Parameters](#unsupported-parameters)
- [Parameters](#parameters)
- [Examples](#examples)
  - [Comments](#comments)
  - [Submissions](#submissions)
- [Advanced Examples](#advanced-examples)
  - [PRAW](#praw)
  - [Custom Filter](#custom-filter)
  - [Caching Examples](#caching-examples)
- [Benchmarks](#benchmarks)
- [Deprecated Examples](#deprecated-examples)

# Description

**PMAW** is a wrapper for the Pushshift API which uses multithreading to retrieve Reddit comments and submissions. General usage is through the `PushshiftAPI` class which provides methods for interacting with different `Pushshift` endpoints, please view the [Pushshift Docs](https://api.pushshift.io/redoc) for more details on the endpoints and accepted parameters. Parameters are provided through keyword arguments when calling the method, some methods will have required parameters. When using a method **PMAW** will complete all the required API calls to complete the query before returning a `Response` generator object.

The following three methods are currently supported:

- Searching Comments: `search_comments`
  - [Details](https://github.com/pushshift/api#searching-comments)
- Search Submissions: `search_submissions`
  - [Details](https://github.com/pushshift/api#searching-submissions)
- Search Submission Comment IDs: `search_submission_comment_ids`
  - [Details](https://github.com/pushshift/api#get-all-comment-ids-for-a-particular-submission)

# Getting Started

## Why Multithread?

When building large datasets from Reddit submission and comment data it can require thousands of API calls to the Pushshift API. The time it takes for your code to complete pulling all this data is limited by both your network latency and the response time of the Pushshift server, which can vary throughout the day.

Current API libraries such as PRAW and PSAW currently run requests sequentially, which can cause thousands of API calls to take many hours to complete. Since API requests are I/O-bound they can benefit from being run asynchronously using multiple threads. Implementing intelligent rate limiting can ensure that we minimize the number of rejected requests, and the time it takes to complete.

## Installation

**PMAW** currently supports _Python 3.5_ or later. To install it via _pip_, run:

```sh
$ pip install pmaw
```

### General Usage

```python
from pmaw import PushshiftAPI
api = PushshiftAPI()
```

View the optional parameters for `PushshiftAPI` [here](#pushshiftapi).

# Features

## Multithreading

The number of threads to use during multithreading is set with the `num_workers` parameter. This is optional and defaults to `10`, however, you should provide a value as this may not be appropriate for your machine. Increasing the number of threads you use allows you to make more concurrent requests to Pushshift, however, the returns are diminishing as requests are constrained by the rate-limit. The optimal number of threads for requests is between `10` and `20` depending on the current response time of the Pushshift server.

When selecting the number of `threads` you can follow one of the two methodologies:

- Number of processors on the machine, multiplied by 5
- Minimum value of 32 and the number of processors plus 4

If you are unsure how many processors you have use: `os.cpu_count()`.

## Rate Limiting

Multiple different options are available for rate-limiting your Pushshift API requests, and are defined by two different types, rate-averaging and exponential backoff. If you're unsure on which to use, refer to the [benchmark comparison](#benchmark-comparison).

### Rate-Averaging

**PMAW** by default rate limits using rate-averaging so that the concurrent API requests to the Pushshift server are limited to your provided rate.

Providing a `rate_limit` value is optional, this defaults to `60` requests per minute which is the recommended value for interacting with the Pushshift API. Increasing this value above `60` will increase the number of rejected requests and will increase the burden on the Pushshift server. A maximum recommended value is `100` requests per minute.

Additionally, the rate-limiting behaviour can be constrained by the `max_sleep` parameter which allows you to select a maximum period of time to sleep between requests.

### Exponential Backoff

Exponential backoff can be used by setting the `limit_type` to backoff. Four flavours of `backoff` are available based on the usage of jitter: None, full, equal, and decorr - decorrelated.

Exponential backoff is calculated by multiplying the `base_backoff` by 2 to the power of the number of failed batches. This allows batches to be spaced out, reducing the resulting rate-limit when requests start to be rejected. However, the threads will still be requesting at nearly the same time, increasing the overall number of required API requests. The exponential backoff sleep values are capped by the `max_sleep` parameter.

Introducing an element of randomness called `jitter` allows us to reduce the competition between threads and distribute the API requests across the window, reducing the number of rejected requests.

- `full` jitter selects the length of sleep for a request by randomly sampling from a normal distribution for values between 0 and the capped exponential backoff value.
- `equal` jitter selects the length of sleep for a request by adding half the capped exponential backoff value to a random sample from a normal distribution between 0 and half the capped exponential backoff value.
- `decorr` - decorrelated jitter is similar to `full` jitter but increases the maximum jitter based on the last random value, selecting the length of sleep by the minimum value between `max_sleep` and a random sample between the `base_backoff` and the last sleep value multiplied by 3.

## Caching

### Memory Safety

Memory safety allows us to reduce the amount of RAM used when requesting data, and can be enabled by setting `mem_safe=True` on a search method. This feature should be used if a large amount of data is being requested or if the machine in use has a limited amount of RAM.

When enabled, **PMAW** caches the responses retrieved every 20 batches (approx 20,000 responses with 10 workers) by default, this can be changed by passing a different value for `file_checkpoint` when instantiating the `PushshiftAPI` object.

When the search is complete, a `Response` generator object is returned, when iterating through the responses using this generator, responses from the cache will be loaded in 1 cache file at a time.

### Safe Exiting

Safe exiting will ensure that if a search method is interrupted that any unfinished requests and current responses are cached before exiting. If the search method successfully completes, all the responses are also cached. This can be enabled by setting `safe_exit=True` on a search method.

Re-running a `search` method with the exact same parameters that you have ran before will load previous responses and any unfinished requests from the cache, allowing it to resume if all the required responses have not yet been retrieved. If there are no unfinished requests, the responses from the cache are returned.

A `before` value is required to load previous responses / requests when using non-id based search, as `before` is set to the current time when the `search` method is called, which would result in a different set of parameters then when you last ran the search despite all other parameters being the same.

Similarly to the memory safety feature, a `Response` generator object is returned. When iterating through the responses using this generator, responses from the cache will be loaded in 1 cache file at a time.

## PRAW Enrichment

Enrich results with the most recent metadata from Reddit by passing a PRAW Reddit instance when instantiating the PushshiftAPI. Results not found on Reddit will not be enriched or returned.

If you don’t already have a client ID and client secret, follow Reddit’s [First Steps Guide](https://github.com/reddit-archive/reddit/wiki/OAuth2-Quick-Start-Example#first-steps) to create them. A user agent is a unique identifier that helps Reddit determine the source of network requests. To use Reddit’s API, you need a unique and descriptive user agent.

## Custom Filtering

A user-defined function can be provided using the `filter_fn` parameter for either the `search_submissions` or `search_comments` method. This function will be used to filter results before they are saved by passing each item to the function and filtering it out if a `False` value is returned, saving the value if `True` is returned. The `limit` parameter does not take into account any results that are filtered out.

## Unsupported Parameters

- `order='asc'` is unsupported as it can have unexpected results
- `until` and `since` only support epoch time (float or int)
- `aggs` are unsupported, as **PMAW** is intended to be used for collecting large numbers of submissions or comments. Use [PSAW](https://github.com/dmarx/psaw) for aggregation requests.

### Feature Requests

- For feature requests please open an issue with the `feature request` label, this will allow features to be better prioritized for future releases

# Parameters

## Objects

### `PushshiftAPI`

- `num_workers` (int, optional): Number of workers to use for multithreading, defaults to 10.
- `max_sleep` (int, optional): Maximum rate-limit sleep time (in seconds) between requests, defaults to 60s.
- `rate_limit` (int, optional): Target number of requests per minute for rate-averaging, defaults to 60 requests per minute.
- `base_backoff` (float, optional): Base delay in seconds for exponential backoff, defaults to 0.5s
- `batch_size` (int, optional): Size of batches for multithreading, defaults to number of workers.
- `shards_down_behavior` (str, optional): Specifies how PMAW will respond if some shards are down during a query. Options are 'warn' to only emit a warning, 'stop' to throw a RuntimeError, or None to take no action. Defaults to 'warn'.
- `limit_type` (str, optional): Type of rate limiting to use, options are 'average' for rate averaging, 'backoff' for exponential backoff. Defaults to 'average'.
- `jitter` (str, optional): Jitter to use with backoff, options are None, 'full', 'equal', 'decorr'. Defaults to None.
- `checkpoint` (int, optional): Size of interval in batches to print a checkpoint with stats, defaults to 10
- `file_checkpoint` (int, optional): Size of interval in batches to cache responses when using mem_safe, defaults to 20
- `praw` (praw.Reddit, optional): Used to enrich the Pushshift items retrieved with metadata directly from Reddit

### `Response`

`Response` is a generator object which will return the responses once when iterated over.

- `len(Response)` will return the number of responses that were retrieved from Pushshift
- `load_cache(key, cache_dir=None)` returns an instance of `Response` with the responses loaded with the provided key

## `search_submissions` and `search_comments`

- `max_ids_per_request` (int, optional): Maximum number of ids to use in a single request, defaults to 500, maximum 500.
- `max_results_per_request` (int, optional): Maximum number of items to return in a single non-id based request, defaults to 100, maximum 100.
- `mem_safe` (boolean, optional): If True, stores responses in cache during operation, defaults to False
- `search_window` (int, optional): Size in days for search window for submissions / comments in non-id based search, defaults to 365
- `safe_exit` (boolean, optional): If True, will safely exit if interrupted by storing current responses and requests in the cache. Will also load previous requests / responses if found in cache, defaults to False
- `cache_dir` (str, optional) - An absolute or relative folder path to cache responses in when `mem_safe` or `safe_exit` is enabled
- `filter_fn` (function, optional) - A function used for custom filtering the results before saving them. Accepts a single comment or submission parameter and returns False to filter out the item, otherwise returns True.

### Keyword Arguments

- Unlike the Pushshift API, the `until` and `since` keyword arguments must be in epoch time
- `limit` is the number of submissions/comments to return. If set to `None` or if the set `limit` is higher than the number of available submissions/comments for the provided parameters then `limit` will be set to the amount available.
- Other accepted parameters are covered in the Pushshift documentation for [submissions](https://github.com/pushshift/api#searching-submissions) and [comments](https://github.com/pushshift/api#searching-comments).

## `search_submission_comment_ids`

- `ids` is a required parameter and should be an array of submission ids, a single id can be passed as a string
- `max_ids_per_request` (int, optional): Maximum number of ids to use in a single request, defaults to 500, maximum 500.
- `mem_safe` (boolean, optional): If True, stores responses in cache during operation, defaults to False
- `safe_exit` (boolean, optional): If True, will safely exit if interrupted by storing current responses and requests in the cache. Will also load previous requests / responses if found in cache, defaults to False
- `cache_dir` (str, optional) - An absolute or relative folder path to cache responses in when `mem_safe` or `safe_exit` is enabled

### Keyword Arguments

- Other accepted parameters are covered in the Pushshift [documentation](https://github.com/pushshift/api#get-all-comment-ids-for-a-particular-submission)

# Examples

The following examples are for `pmaw` version >= 1.0.0.

## Comments

### Search Comments

```python
from pmaw import PushshiftAPI

api = PushshiftAPI()
comments = api.search_comments(subreddit="science", limit=1000)
comment_list = [comment for comment in comments]
```

### Search Comments by IDs

```python
from pmaw import PushshiftAPI

api = PushshiftAPI()
comment_ids = ['gjacwx5','gjad2l6','gjadatw','gjadc7w','gjadcwh',
  'gjadgd7','gjadlbc','gjadnoc','gjadog1','gjadphb']
comments = api.search_comments(ids=comment_ids)
comment_list = [comment for comment in comments]
```

You can supply a single comment by passing the id as a string or an array with a length of 1 to `ids`

[Detailed Example](https://github.com/mattpodolak/pmaw/blob/master/examples/search_comments.ipynb)

### Search Comment IDs by Submission ID

```python
from pmaw import PushshiftAPI

api = PushshiftAPI()
post_ids = ['kxi2w8','kxi2g1','kxhzrl','kxhyh6','kxhwh0',
  'kxhv53','kxhm7b','kxhm3s','kxhg37','kxhak9']
comment_ids = api.search_submission_comment_ids(ids=post_ids)
comment_id_list = [c_id for c_id in comment_ids]
```

You can supply a single submission by passing the id as a string or an array with a length of 1 to `ids`

[Detailed Example](https://github.com/mattpodolak/pmaw/blob/master/examples/search_submission_comment_ids.ipynb)

## Submissions

### Search Submissions

```python
from pmaw import PushshiftAPI

api = PushshiftAPI()
posts = api.search_submissions(subreddit="science", limit=1000)
post_list = [post for post in posts]
```

### Search Submissions by IDs

```python
from pmaw import PushshiftAPI

api = PushshiftAPI()
post_ids = ['kxi2w8','kxi2g1','kxhzrl','kxhyh6','kxhwh0',
  'kxhv53','kxhm7b','kxhm3s','kxhg37','kxhak9']
posts = api.search_submissions(ids=post_ids)
post_list = [post for post in posts]
```

You can supply a single submission by passing the id as a string or an array with a length of 1 to `ids`

[Detailed Example](https://github.com/mattpodolak/pmaw/blob/master/examples/search_submissions.ipynb)

# Advanced Examples

## PRAW

```python
import praw
from pmaw import PushshiftAPI

reddit = praw.Reddit(
 client_id='YOUR_CLIENT_ID',
 client_secret='YOUR_CLIENT_SECRET',
 user_agent=f'python: PMAW request enrichment (by u/YOUR_USERNAME)'
)

api_praw = PushshiftAPI(praw=reddit)
comments = api_praw.search_comments(q="quantum", subreddit="science", limit=100, until=1629990795)
```

## Custom Filter

The user defined function must accept a single item (comment / submission) and return either True or False, returning False will filter out the item passed to it.

```python
from pmaw import PushshiftAPI

api = PushshiftAPI()
def fxn(item):
  return item['score'] > 2
posts = api.search_submissions(ids=post_ids, filter_fn=fxn)
```

## Caching Examples

### Memory Safety

If you are pulling large amounts of data or have a limited amount of RAM, using the memory safety feature will help you avoid an out of memory error from being thrown during data retrieval.

```python
from pmaw import PushshiftAPI

api = PushshiftAPI()
posts = api.search_submissions(subreddit="science", limit=700000, mem_safe=True)
print(f'{len(posts)} posts retrieved from Pushshift')
```

A `Response` generator object will be returned, and you can load all the responses, including those that have been cached by iterating over the entire generator.

```python
# get all responses
post_list = [post for post in posts]
```

With default settings, responses are cached every 20 batches (approx 20,000 responses with 10 workers), however, with limited memory you can decrease this further.

```python
# cache responses every 10 batches
api = PushshiftAPI(file_checkpoint=10)
```

### Safe Exiting

If you expect that your query may be interrupted while its running, setting `safe_exit=True` will cache responses and unfinished requests before exiting when an interrupt signal is received. Re-running a `search` method with the exact same parameters that you have ran before will load previous responses and any unfinished requests from the cache, allowing it to resume if all the required responses have not yet been retrieved.

```python
from pmaw import PushshiftAPI

api = PushshiftAPI()
posts = api.search_submissions(subreddit="science", limit=700000, until=1613234822, safe_exit=True)
print(f'{len(posts)} posts retrieved from Pushshift')
```

A `until` value is required to load previous responses / requests when using non-id search, as `until` is set to the current time when the `search` method is called, which would result in a different set of parameters then when you last ran the search despite all other parameters being the same.

### Loading Cache with Key

The `Response` class has a staticmethod `load_cache` that allows you to pass a `key` value and `cache_dir` to load responses from the cache into an instance of `Response`.

```python
from pmaw import Response

cache_key = 'a904e22ea572a8c0109b7bc5330528e4'
cache_dir = './cache'
resp = Response.load_cache(cache_key, cache_dir)

```

# Benchmarks

[Benchmark Notebook](https://github.com/mattpodolak/pmaw/blob/master/examples/benchmark.ipynb)

## PMAW and PSAW Comparison

### Completion Time

A benchmark comparison was performed to determined the completion time for different size requests, ranging from 1 to 390,000 requested posts. This will allow us to determine which Pushshift wrappers and rate-limiting methods are best for different request sizes.

Default parameters were used for each **PMAW** rate-limit configuration as well as the default **PSAW** configuration, which does not provide multiple rate-limit implementations.

![01 benchmark](https://raw.githubusercontent.com/mattpodolak/pmaw/master/examples/img/01-comparison.png)

For the first benchmark test we compare the completion times for all possible **PMAW** rate-limiting configurations with **PSAW** for up to 16,000 requested posts. We can see that the three most performant rate-limiting settings for **PMAW** are rate-averaging, and exponential backoff with full or equal jitter.

![02 benchmark](https://raw.githubusercontent.com/mattpodolak/pmaw/master/examples/img/02-comparison.png)

We ran this second benchmark increasing up to 390,000 requested posts, excluding the least performant **PMAW** rate-limiting configurations. From this benchmark, we can see that **PMAW** was on average 1.79x faster than **PSAW** at 390,625 posts retrieved. The total completion time for 390,625 posts with **PSAW** was 2h38m, while the average completion time was 1h28m for **PMAW**.

### Number of Requests


CHANGES.md
## 3.0.0 (2022/12/24)

- changed `before` and `after` to `until` and `since`
- removed `metadata=true` as this is now always enabled
- set `order='desc'` as this replaces `sort`
- set `sort='created_utc'` so that slicing still works as expected
- Read more on [COLO switchover](https://www.reddit.com/r/pushshift/comments/zkggt0/update_on_colo_switchover_bug_fixes_reindexing/)
- refactored metadata usage
- 🎅🎅🎅🎅🎅🎅🎅🎅🎅🎅🎅🎅🎅🎅

## 2.1.3 (2022/02/20)

- Don't inherit from object in classes
- Removed logging configuration to prevent unexpected results for users

## 2.1.2 (2022/01/07)

- fix scenario where a result is reported but cannot be returned by Pushshift

## 2.1.1 (2021/11/29)

- fix index error bug

## 2.1.0 (2021/10/01)

- Updated logging and set default log level to INFO
- Added `load_cache` static method to `Response` to load cached responses using cache key

## 2.0.0 (2021/09/11)

- Added support for enriching result metadata using PRAW
- Implemented functional tests
- Reduced `max_ids_per_request` to 500
- Added automated testing
- Increased exception handling specificity
- Added `filter_fn` for custom filtering

## 1.1.0 (2021/05/27)

- Added gzip for cached pickle files
- Exception handling is now slightly more specific
- Updated many print statements to output via logging
- Fixed issue with safe_exit not saving info

## 1.0.5 (2021/04/21)

- Moved remaining limit logging to DEBUG from INFO
- Fixed generator incorrect length after being partially iterated through
- Reduced the number of debug logs
- Fixed duplicate responses being returned if the number of responses for a provided window is less than expected

## 1.0.4 (2021/03/05)

- None type comparison bug fixed
- updated how limit was being updated for submission comment ids

## 1.0.3 (2021/02/19)

- fixed early cache bug
- fixed limit being retrieved from next search window when resuming from safe exit

## 1.0.2 (2021/02/16)

- fixed comments returning 25 by default

## 1.0.1 (2021/02/16)

- limit error in `trim` hot fix

## 1.0.0 (2021/02/14)

- `search` methods now return a `Response` generator object
- memory safety can now be enabled with `mem_safe` to cache responses during data retrieval and reduce the amount of memory used
- safe exiting can now be enabled with `safe_exit` to safely exit when an interrupt signal is received during data retrieval
- load unfinished requests and saved responses from `cache` when safe exiting is enabled
- request details are now handled inside a `Request` object

## 0.1.3 (2021/02/08)

- Fixed infinite while loop error
- Checkpoint by batch
- Removed erroneous pandas import

## 0.1.2 (2021/02/06)

- Fixed timeslicing creating extra requests

## 0.1.1 (2021/02/06)

- Fixed a bug with timeslicing causing duplicate results
- Fixed a miscalculation error for remaining results for a timeslice

## 0.1.0 (2021/02/05)

- General code improvements
- Added exponential backoff and jitter rate-limiting
- Added `non-id` search for submissions and comments

## 0.0.2 (2021/01/23)

- Initial implementation of multithreading requests for `ids` queries, with support for:
  - comment ids by submission id
  - submissions by id
  - comments by id
- Rate-limit based on rate averaging across previous requests


pmaw/PushshiftAPIBase.py
import json
import copy
import logging
from concurrent.futures import ThreadPoolExecutor, as_completed

import requests
from pmaw.types.exceptions import HTTPError, HTTPNotFoundError
from pmaw.Metadata import Metadata

from pmaw.RateLimit import RateLimit
from pmaw.Request import Request


log = logging.getLogger(__name__)


class PushshiftAPIBase:
    _base_url = "https://{domain}.pushshift.io/{{endpoint}}"

    def __init__(
        self,
        num_workers=10,
        max_sleep=60,
        rate_limit=60,
        base_backoff=0.5,
        batch_size=None,
        shards_down_behavior="warn",
        limit_type="average",
        jitter=None,
        checkpoint=10,
        file_checkpoint=20,
        praw=None,
    ):
        self.num_workers = num_workers
        self.domain = "api"
        self.shards_down_behavior = shards_down_behavior
        self.meta = Metadata({})
        self.resp_dict = {}
        self.checkpoint = checkpoint
        self.file_checkpoint = file_checkpoint
        self.praw = praw

        if batch_size:
            self.batch_size = batch_size
        else:
            self.batch_size = num_workers

        # instantiate rate limiter
        self._rate_limit = RateLimit(
            rate_limit, base_backoff, limit_type, max_sleep, jitter
        )

    @property
    def base_url(self):
        # getter for base_url, with formatted domain
        return self._base_url.format(domain=self.domain)

    def _impose_rate_limit(self):
        interval = self._rate_limit.delay()
        if interval > 0:
            self.req._idle_task(interval)

    def _get(self, url, payload={}):
        self._impose_rate_limit()
        r = requests.get(url, params=payload)
        status = r.status_code
        reason = r.reason

        if status == 200:
            r = json.loads(r.text)

            # check if shards are down
            self.meta = Metadata(r.get("metadata", {}))
            total_results = self.meta.total_results
            if total_results:
                after, before = self.meta.ranges
                if after and before:
                    self.resp_dict[(after, before)] = total_results

            return r["data"]
        else:
            if status == 404:
                raise HTTPNotFoundError(f"HTTP {status} - {reason}")
            else:
                # TODO: add custom error types for rate limit and other errors
                raise HTTPError(f"HTTP {status} - {reason}")

    def _multithread(self, check_total=False):
        with ThreadPoolExecutor(max_workers=self.num_workers) as executor:

            while len(self.req.req_list) > 0 and not self.req.exit.is_set():
                # reset resp_dict which tracks remaining responses for timeslices
                self.resp_dict = {}

                # set number of futures created to batch size
                reqs = []
                if check_total:
                    reqs.append(self.req.req_list.popleft())
                else:
                    for i in range(min(len(self.req.req_list), self.batch_size)):
                        reqs.append(self.req.req_list.popleft())

                futures = {
                    executor.submit(self._get, url_pay[0], url_pay[1]): url_pay
                    for url_pay in reqs
                }

                self._futures_handler(futures, check_total)

                # reset attempts if no failures
                self._rate_limit._check_fail()

                # check if shards are down
                if self.meta.shards_are_down and (
                    self.shards_down_behavior is not None
                ):
                    shards_down_message = "Not all PushShift shards are active. Query results may be incomplete."
                    if self.shards_down_behavior == "warn":
                        log.warning(shards_down_message)
                    if self.shards_down_behavior == "stop":
                        self._shutdown(executor)
                        raise RuntimeError(
                            shards_down_message
                            + f" {len(self.req.req_list)} unfinished requests."
                        )
                if not check_total:
                    self.num_batches += 1
                    if self.num_batches % self.file_checkpoint == 0:
                        # cache current results
                        executor.submit(self.req.save_cache())
                    self._print_stats("Checkpoint")
                else:
                    break
            if not check_total:
                self._print_stats("Total")
            self._shutdown(executor)

    def _futures_handler(self, futures, check_total):
        for future in as_completed(futures):
            url_pay = futures[future]
            self.num_req += int(not check_total)
            try:
                data = future.result()
                self.num_suc += int(not check_total)
                url = url_pay[0]
                payload = url_pay[1]
                if not check_total:
                    self.req.save_resp(data)

                    log.debug(f"Remaining limit {self.req.limit}")
                    if self.req.limit <= 0:
                        log.debug(
                            f"Cancelling {len(self.req.req_list)} unfinished requests"
                        )
                        self.req.req_list.clear()
                        break

                    # handle time slicing logic
                    if "until" in payload and "since" in payload:
                        until = payload["until"]
                        since = payload["since"]
                        log.debug(
                            f"Time slice from {since} - {until} returned {len(data)} results"
                        )
                        total_results = self.resp_dict.get((since, until), 0)
                        log.debug(f"{total_results} total results for this time slice")
                        # calculate remaining results
                        remaining = total_results - len(data)

                        # number of timeslices is depending on remaining results
                        if remaining > self.req.max_results_per_request * 2:
                            num = 2
                        elif remaining > 0:
                            num = 1
                        else:
                            num = 0

                        if num > 0:
                            # find minimum `created_utc` to set as the `before` parameter in next timeslices
                            # Fix issue where Pushshift occasionally reports remaining results that it is
                            # unable to return - len(data) == 0 when this happens
                            if len(data) > 0:
                                until = data[-1]["created_utc"]
                                # generate payloads
                                self.req.gen_slices(url, payload, since, until, num)

            except HTTPNotFoundError as exc:
                log.debug(f"Request Failed -- {exc}")
                # dont retry ids not found
                # it looks like submission/comment_ids/ returns 404s now
                if "ids" not in self.req.payload:
                    self.req.req_list.appendleft(url_pay)

            except HTTPError as exc:
                log.debug(f"Request Failed -- {exc}")
                self._rate_limit._req_fail()
                self.req.req_list.appendleft(url_pay)

    def _shutdown(self, exc, wait=False, cancel_futures=True):
        # shutdown executor
        try:
            # pass cancel_futures keywords avail in python 3.9
            exc.shutdown(wait=wait, cancel_futures=cancel_futures)
        except TypeError:
            # TODO: manually cancel pending futures
            exc.shutdown(wait=wait)

    def _print_stats(self, prefix):
        rate = self.num_suc / self.num_req * 100
        remaining = self.req.limit
        if (self.num_batches % self.checkpoint == 0) and prefix == "Checkpoint":
            log.info(
                f"{prefix}:: Success Rate: {rate:.2f}% - Requests: {self.num_req} - Batches: {self.num_batches} - Items Remaining: {remaining}"
            )
        elif prefix == "Total":
            if remaining < 0:
                remaining = 0  # don't print a neg number
            log.info(
                f"{prefix}:: Success Rate: {rate:.2f}% - Requests: {self.num_req} - Batches: {self.num_batches} - Items Remaining: {remaining}"
            )
            if self.req.praw and len(self.req.enrich_list) > 0:
                # let the user know praw enrichment is still in progress so it doesnt appear to hang after
                # finishing retrieval from Pushshift
                log.info(f"Finishing enrichment for {len(self.req.enrich_list)} items")

    def _reset(self):
        self.num_suc = 0
        self.num_req = 0
        self.num_batches = 0

    def _search(
        self,
        kind,
        max_ids_per_request=500,
        max_results_per_request=100,
        mem_safe=False,
        search_window=365,
        dataset="reddit",
        safe_exit=False,
        cache_dir=None,
        filter_fn=None,
        **kwargs,
    ):

        # TODO: remove this warning once 404s stop happening
        if kind == "submission_comment_ids":
            log.warning(
                "submission comment id search may return no results due to COLO switchover"
            )

        # raise error if aggs are requested
        if "aggs" in kwargs:
            err_msg = "Aggregations support for {} has not yet been implemented, please use the PSAW package for your request"
            raise NotImplementedError(err_msg.format(kwargs["aggs"]))

        self.meta = Metadata({})
        self.resp_dict = {}
        self.req = Request(
            copy.deepcopy(kwargs),
            filter_fn,
            kind,
            max_results_per_request,
            max_ids_per_request,
            mem_safe,
            safe_exit,
            cache_dir,
            self.praw,
        )

        # reset stat tracking
        self._reset()

        if kind == "submission_comment_ids":
            endpoint = f"{dataset}/submission/comment_ids/"
        else:
            endpoint = f"{dataset}/{kind}/search"

        url = self.base_url.format(endpoint=endpoint)

        while (
            self.req.limit is None or self.req.limit > 0
        ) and not self.req.exit.is_set():
            # set/update limit
            if "ids" not in self.req.payload and len(self.req.req_list) == 0:
                # check to see how many results are remaining
                self.req.req_list.appendleft((url, self.req.payload))
                self._multithread(check_total=True)
                total_avail = self.meta.total_results

                if self.req.limit is None:
                    log.info(f"{total_avail} result(s) available in Pushshift")
                    self.req.limit = total_avail
                elif total_avail < self.req.limit:
                    log.info(
                        f"{self.req.limit - total_avail} result(s) not found in Pushshift"
                    )
                    log.info(f"{total_avail} total available")
                    self.req.limit = total_avail

            # generate payloads
            self.req.gen_url_payloads(url, self.batch_size, search_window)

            # check for exit signals
            self.req.check_sigs()

            if self.req.limit > 0 and len(self.req.req_list) > 0:
                self._multithread()

        self.req.save_cache()
        return self.req.resp
